palace/fem/libceed/operator.cpp

```diff
@@ -508,17 +487,23 @@ std::unique_ptr<Operator> CeedOperatorCoarsen(const Operator &op_fine,
                                                        fespace_coarse.GetVSize());
 
   // Assemble the coarse operator by coarsening each sub-operator (over threads, geometry
-  // types, integrators) of the original fine operator. We loop over Ceed contexts because
-  // extracting the Ceed context from a CeedOperator returns a different pointer (created
-  // with CeedReferenceCopy) and we need the original ones to access the FiniteElementSpace
-  // and Mesh object caches.
+  // types, integrators) of the original fine operator.
   MFEM_VERIFY(internal::GetCeedObjects().size() == op_fine.Size(),
               "Unexpected size mismatch in multithreaded Ceed contexts!");
-  const std::size_t nt = internal::GetCeedObjects().size();
+  const std::size_t nt = op_fine.Size();
   PalacePragmaOmp(parallel for schedule(static))
   for (std::size_t i = 0; i < nt; i++)
   {
-    Ceed ceed = internal::GetCeedObjects()[i];
+    Ceed ceed;
+    PalaceCeedCallBackend(CeedOperatorGetCeed(op_fine[i], &ceed));
+    {
+      Ceed ceed_parent;
+      PalaceCeedCall(ceed, CeedGetParent(ceed, &ceed_parent));
+      if (ceed_parent)
+      {
+        ceed = ceed_parent;
+      }
```

Comment:
This has to do with how libCEED handles backend overrides for certain functionality. Certain backends only implement some functionality (for example, `/gpu/cuda/magma` only implements `CeedBasis` but falls back to a base implementation of `CeedElemRestriction`). Thus, the `Ceed` context attached to the given operator object and returned with `CeedOperatorGetCeed` might not match the parent `Ceed` which Palace knows about because internally the functionality has been delegated to a "base" `Ceed` context. From Palace we cache objects based on the parent `Ceed` context object and thus we need to make sure here we're getting the right context via the parent, otherwise the calls in `SingleOperatorCoarsen` will not work.